staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+		// to be requeued, and hopefully by the time it's processed again the inconsistency has been resolved
+		err = fmt.Errorf("Att %s has inconsistent state: found both in local atts cache and remote atts cache for VNI %d",
+			attNSN,
+			vni)
+	case attAsLocal != nil && attAsRemote == nil:
+		// If we're here the attachment was found only in the local cache: that's the univocal version of the attachment
+		att = attAsLocal
+	case attAsLocal == nil && attAsRemote != nil:
+		// If we're here the attachment was found only in the remote attachments cache for its vni: that's the univocal version of the attachment
+		att = attAsRemote
+	default:
+		// If we're here neither lookup could find the attachment: we assume the attachment has been
+		// deleted by both caches and is therefore no longer relevant to the connection agent
+		attDeleted = true
+	}
+	return
+}
+
+func (ca *ConnectionAgent) processExistingAtt(att *netv1a1.NetworkAttachment) error {
+	attNSN, attVNI, attNode := kosctlrutils.AttNSN(att), att.Spec.VNI, att.Spec.Node
+
+	// Retrieve the last seen attachment state, which stores the attachment network
+	// interface if it was created, and the vni with which it was processed last time
+	// (this field is used to update the old vnState in case the current version of the
+	// attachment has a different vni).
+	attState := ca.getAttState(attNSN)
+	if attState == nil {
+		attState = &attachmentState{
+			vnStateVNI: attVNI,
+		}
+		ca.setAttState(attNSN, attState)
+	}
+
+	// Update the vnState associated with the attachment. This typically involves adding
+	// the attachment to the vnState associated to its vni (and initializing that vnState
+	// if the attachment is the first local one with its vni), but could also entail removing
+	// the attachment from the vnState associated with its old vni if the vni has changed.
+	var err error
+	vnState, firstLocalAttInVN := ca.updateVNState(attState, attVNI, attNSN, attNode)
+	if firstLocalAttInVN {
+		glog.V(2).Infof("VN with ID %d became relevant: an Informer has been started", attVNI)
+	}
+	if vnState != nil {
+		// If we're here att is currently remote but was previously the last local attachment
+		// in its vni. Thus, we act as if the last local attachment in the vn was deleted
+		ca.stopRemoteAttInformerAndEnqueueRemoteAttRefs(vnState, attVNI)
+		return nil
+	}
+
+	// Update attState with the most recent vni.
+	attState.vnStateVNI = attVNI
+
+	// Create or update the interface associated with the attachment.
+	attGuestIP := gonet.ParseIP(att.Status.IPv4)
+	var attHostIP gonet.IP
+	if attNode == ca.localNodeName {
+		attHostIP = ca.hostIP
+	} else {
+		attHostIP = gonet.ParseIP(att.Status.HostIP)
+	}
+	attIfc, attHasIfc, err := ca.createOrUpdateIfc(attState, attGuestIP, attHostIP, attVNI, attNSN)
+
+	// Update attState with the most recent interface (if it's been created)
+	attState.ifc, attState.ifcIsValid = attIfc, attHasIfc
```

Comment:
Why do this update here instead of in `ca.createOrUpdateIfc` ?

Reply:
When I wrote this I wanted attState to be modified only by high-level methods, while being read-only for lower level ones.